### ISSUE TITLE
Trouble embedding bitcode into object file

### ISSUE THREAD

**Issue by hdwhittaker** [mull-project/mull#869]
Hello,

I'm trying to embed bitcode into an executable for the MQTT protocol library and I'm running into some problems. I compile a lot of files into .bc files using a line similar to this one:

`clang -I. -I.. -I../lib -I../src/deps -Wall -ggdb -c -O0 -emit-llvm -fno-inline -DVERSION="\"1.6.7\"" -DWITH_BROKER -fembed-bitcode -c mosquitto.c -o mosquitto.bc`

and then I link all the files together using `llvm-link-10`. Once it links and I try to run mull-cxx on the resulting executable, it says **"The file was not recognized as a valid object file"** and then no mutants were found. So it looks like it is able to embed the bitcode into the executable, but it isn't creating a valid object file for some reason. It's a lot of files so I'm assuming that _some_ mutant would be generated somewhere, I'm curious why it looks like none can be found. My guess guess is it has to do with some way I'm compiling the executable. Does anyone have any ideas on how I can fix this? Any help would be greatly appreciated!

**Comment by AlexDenisov** [comment]
Hi @hdwhittaker!
If I understand correctly, then you don't need to emit LLVM bitcode explicitly (`-emit-llvm`) and you also don't need to link the bitcode files manually.
In other words, simply adding `-fembed-bitcode` to the compiler/linker flags should do the job just fine.

Are you compiling MQTT manually or are you using its own build system?

**Comment by hdwhittaker** [comment]
Hi @AlexDenisov!

Thanks for getting back to me! I removed that flag during compiling and the linker is returning "error: expected top-level entity" then it prints a lot of gibberish. I'm using a Makefile that compiles all of the .c files of MQTT into .bc files and then the Makefile links all of the .bc files together using the llvm linker. 

Using the method I had in place, the executable that was created was running properly and Mull was able to run and extract the bitcode, however it said no mutants were found. My assumption was when Mull said "not a valid object file", that was causing the lack of found mutants/why Mull couldn't do its thing. Do you think that's the problem or would it likely be something else?

Thanks, 
hdwhittaker

**Comment by AlexDenisov** [comment]
@hdwhittaker could you share more details (e.g. mull output, the build script, etc.)?
I don't understand how you can get a normal executable using llvm linker: AFAIK, llvm-link only merges a bunch of LLVM modules into big module.
At which point does mull print "not a valid object file" error message?

**Comment by hdwhittaker** [comment]
@AlexDenisov Sure thing.

**Build script:** The Makefile builds a lot of targets `OBJS = mosquitto.bc \ bridge.bc \ conf.bc ...` that are included in the mosquitto project. Each .bc file is built in the way I mentioned before `clang -I. -I.. -I../lib -I../src/deps -Wall -ggdb -c -O0 -emit-llvm -fno-inline -DVERSION="\"1.6.7\"" -DWITH_BROKER -fembed-bitcode -c mosquitto.c -o mosquitto.bc` where instead of `mosquitto.c` and `mosquitto.bc`, it would be each included file and the target .bc file (Note that I readded the -emit-llvm flag because I was getting an error). Once all of those files are built, they are linked together with the command `llvm-link-10 ${OBJS} -o mosquitto` and `mosquitto` would be the executable that is created. The executable runs just fine. 

**Mull Output:** After executing the command `mull-cxx mosquitto`, the output is as follows:
![temp](https://user-images.githubusercontent.com/77063162/120823238-3c3ca000-c525-11eb-82f3-4ccf81535ea7.PNG)

**Mosquitto:** When I run mosquitto, I enter `mosquitto` into the command line and this is what comes up:
![temp1](https://user-images.githubusercontent.com/77063162/120823982-fa602980-c525-11eb-92a8-48f78f2d1981.PNG)

**Comment by AlexDenisov** [comment]
@hdwhittaker thank you very much for sharing, I didn't know it's possible to do it the way you did 😄.
I'll run some experiments and will come back to you soon.

**Comment by hdwhittaker** [comment]
@AlexDenisov Okay! Please let me know if you get any luck!

**Comment by AlexDenisov** [comment]
@hdwhittaker I'm trying a smaller example and failing to reproduce the issue:

```
$ cat main.c
int main() {
  return 0;
}
$ clang-10 -c -O0 -emit-llvm -fembed-bitcode main.c -o main.bc
$ llvm-link-10 main.bc -o main
$ file main.bc
main.bc: LLVM IR bitcode
$ file main
main: LLVM IR bitcode
$ mull-cxx ./main
[info] Extracting bitcode from executable (threads: 1)
The file was not recognized as a valid object file
       [################################] 1/1. Finished in 0ms
[info] Sanity check run (threads: 1)
[error] Cannot run executable: Permission denied
stdout:
stderr:

[error] Error messages are treated as fatal errors. Exiting now.
```

What I certainly do not understand is how you can get an executable with just `llvm-link`.

---

On a slightly different, but still related topic: shall we try to embed bitcode without intermediate `.bc` files and without `llvm-link`? I.e., shall we debug the issues you are having with the following approach https://mull.readthedocs.io/en/latest/HelloWorld.html#step-2-enabling-bitcode ?

**Comment by hdwhittaker** [comment]
@AlexDenisov Hmm this is interesting. I created a file identical to what you have in `main.c` and I created an object file the way it is in the Hello World example. It runs properly if I use `./`, but it fails if I omit it. The opposite seems to be happening with what I have currently.

I re-tried running mull against the `mosquitto` linked file I generated and it varies depending on how I run it. If I do something similar to what you did, `mull-cxx ./mosquitto` it fails to run and looks similar to the output you have (Cannot run executable: Permission denied), however the way I was doing it before, `mull-cxx mosquitto`, it looks like mull is able to complete. 

Any idea why/how this is happening?

**Comment by hdwhittaker** [comment]
Hey @AlexDenisov 

I tried recreating the Hello World example in the documentation and I noticed a few things. Mull seemed to return mutation scores that didn't reflect what I thought it would return when I altered a few things. For example, when I include the clang flag `-O0` (baseline optimization, which is essentially no optimizations from what I can tell), Mull would bizarrely kill mutants when I didn't expect it to (things like killing a >= operator in ageGreaterThan21(23) with < and > operators). Additionally, if I take some small liberties with how the tests are executed, Mull would somehow kill mutants that shouldn't be killed. For example, if I use `if(!ageGreaterThan21(23))` versus `if(ageGreaterThan21(23) != true)`, the mutation score would be incorrectly different (similar to what I mentioned earlier) even though they seem to be functionally equivalent. Does Mull catch expressions in `if` statements or does Mull only catch expressions that are of the form `bool value = foo == bar`? If you need me to post my code so you can better troubleshoot it, I can.

**Comment by AlexDenisov** [comment]
Hi @hdwhittaker, I just tried reproducing what you described and could not see the problem.
Sharing what you have (exact code and steps you take) would be of great help in debugging this issue.

**Comment by hdwhittaker** [comment]
The following code when compiled with `clang -c -O0 -fembed-bitcode -g temp.c -o temp` will occasionally return mutation scores of 50% as well as 100% without any edits.
```
#include <stdbool.h>

bool age(int input){
   if (input >= 21){ return true; }
   else { return false; }
}

int main(){
   if (!age(23)){
      return 1;
   }
   if (age(19)){
      return 1;
   }
   return 0;
}
```
For the 50% score, the surviving mutant would be `Replaced >= with >` for the `if (input >= 21){ return true; }` line. The 100% mutation score would kill that same exact mutant.

Any idea why this could be happening?

**Comment by AlexDenisov** [comment]
It's an interesting example. On my machine (macOS) it constantly gives 50%. I'll give it more tries on Linux to see if I can reproduce it.
So far, I don't have any ideas about what could go wrong in this case.

**Comment by AlexDenisov** [comment]
Hi @hdwhittaker, I hope the issue is resolved via #891, could you give it another try using a specific version as outlined here https://github.com/mull-project/mull/issues/890#issuecomment-892138922?

**Comment by AlexDenisov** [comment]
Hi @hdwhittaker, consider giving a try to the new IR frontend. Here is the tutorial
https://mull.readthedocs.io/en/0.15.0/tutorials/HelloWorld.html
